tephra-core/src/main/java/co/cask/tephra/visibility/VisibilityFence.java

```diff
@@ -0,0 +1,128 @@
+ * If the writer is able to successfully commit the transaction with write fence, it means that
+ * any in-progress readers that have not seen the change will not be allowed to commit anymore.
+ * In case an in-progress reader commits before the write fence is committed, then
+ * the writer will get a conflict on commit, and the writer has to retry the write fence with a new transaction
+ * until it successfully commits.
+ * <p>
+ * Any in-progress readers after the write fence is committed will get a conflict when they try to commit. This will
+ * force the reader to start a new transaction, and this ensures that the changes made by writer are visible
+ * to the reader.
+ * <p>
+ * Hence a successful commit of a write fence ensures that any read fence that successfully commits after
+ * this point onwards would have seen the change.
+ *
+ * <p>
+ * Sample reader code:
+ * <pre>
+ *   <code>
+ * TransactionAware readFence = VisibilityFence.reader(fenceId);
+ * TransactionContext readTxContext =
+ *        new TransactionContext(new InMemoryTxSystemClient(txManager), readFence, table1, table2, ...);
+ * readTxContext.start();
+ *
+ * // do operations using table1, table2, etc.
+ *
+ * // finally commit
+ * try {
+ *   readTxContext.finish();
+ * } catch (TransactionConflictException e) {
+ *   // handle conflict by aborting and starting over with a new transaction using the same fence object
+ * }
+ *   </code>
+ * </pre>
+ *<p>
+ *
+ * Sample writer code:
+ * <pre>
+ *   <code>
+ * // start transaction
+ * // write change
+ * // commit transaction
+ *
+ * // Now use a write fence (with the same fenceId) to ensure that all in-progress readers are aware of this change
+ * TransactionAware writeFence = VisibilityFence.writer(fenceId);
+ * TransactionContext writeTxContext = new TransactionContext(new InMemoryTxSystemClient(txManager), writeFence);
+ * writeTxContext.start();
+ * try {
+ *   writeTxContext.finish();
+ * } catch (TransactionConflictException e) {
+ *   // handle conflict by aborting and starting over with a new transaction using the same fence object
+ * }
+ *   </code>
+ * </pre>
+ *
+ * fenceId in the above samples refers to any id that both the readers and writer know for a given
+ * piece of data. Both readers and writer will have to use the same fenceId to synchronize on a given data.
+ * Typically fenceId uniquely identifies the data in question.
+ * For example, if the data is a table row, the fenceId can be
+ * composed of table name and row key. If the data is a table cell, the fenceId can be composed of
+ * table name, row key, and column key.
+ *<p>
+ *
+ * Note that in this implementation, any reader that starts a transaction after the write is committed,
+ * but before the transaction with write fence is started will also conflict with the write fence,
+ * even though this reader has seen the latest changes.
+ */
+public class VisibilityFence {
+  private static final byte[] VISIBILITY_PREFIX = {'v'};
```

Comment:
Why need the prefix? Shouldn't the participants just pick a unique fenceId as a prefix that won't clash with their change sets?